pkg/apis/kubeone/validation/validation.go

```diff
@@ -352,17 +352,45 @@ func ValidateContainerRuntimeConfig(cr kubeoneapi.ContainerRuntimeConfig, versio
 func ValidateClusterNetworkConfig(c kubeoneapi.ClusterNetworkConfig, fldPath *field.Path) field.ErrorList {
 	allErrs := field.ErrorList{}
 
-	if len(c.PodSubnet) > 0 {
-		if _, _, err := net.ParseCIDR(c.PodSubnet); err != nil {
-			allErrs = append(allErrs, field.Invalid(fldPath.Child("podSubnet"), c.PodSubnet, ".clusterNetwork.podSubnet must be a valid CIDR string"))
-		}
+	subnetCountErr := func(node, subnet string, count int, ipFamily kubeoneapi.IPFamily) *field.Error {
+		return field.Invalid(fldPath.Child(node), subnet, fmt.Sprintf(".clusterNetwork.%s must specify %d subnet(s) for %q IP family.", node, count, ipFamily))
+	}
+
+	invalidFamilyErr := func(node, subnet string, ipFamily kubeoneapi.IPFamily) *field.Error {
+		return field.Invalid(fldPath.Child(node), subnet, fmt.Sprintf(".clusterNetwork.%s must be valid %q subnet.", node, ipFamily))
 	}
-	if len(c.ServiceSubnet) > 0 {
-		if _, _, err := net.ParseCIDR(c.ServiceSubnet); err != nil {
-			allErrs = append(allErrs, field.Invalid(fldPath.Child("serviceSubnet"), c.ServiceSubnet, ".clusterNetwork.serviceSubnet must be a valid CIDR string"))
+
+	var subnetValidators []func(string) bool
+	switch c.IPFamily {
+	case kubeoneapi.Unspecified, kubeoneapi.IPv4:
+		subnetValidators = append(subnetValidators, netutils.IsIPv4CIDRString)
+	case kubeoneapi.IPv6:
+		subnetValidators = append(subnetValidators, netutils.IsIPv6CIDRString)
+	case kubeoneapi.DualStack:
+		subnetValidators = append(subnetValidators, netutils.IsIPv4CIDRString, netutils.IsIPv6CIDRString)
+	default:
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("ipFamily"), c.IPFamily, "unknown ipFamily"))
+	}
+
+	validateCIDRs := func(node, subnet string) {
+		if len(subnet) == 0 {
+			return
+		}
```

Comment:
Shouldn't this be an invalid case? I guess kubeadm will default to something, but I'd rather be explicit here.

Reply:
Will check defaulting. I didn't add any defaulting so we might be ending up here.